### DISCUSSION TITLE
Keeping preamble as AST

### DISCUSSION THREAD

**Discussion by kuganv** [clangd/clangd#1240] [Ideas]
clangd creates PCH as a serialised AST which clang understands. PCH can also be stored on disk which potentially saves memory. However it also results in performance hit due to the time taken to serialise and de-serialise the AST. Has anyone experimented with keeping the preamble as AST in memory and making clan use this build AST for the TU. 

Thanks, Kugan

**Comment by Trass3r** [comment]
You mean clang re-using clangd's, not [this](https://github.com/llvm/llvm-project/blob/defe072010ac365981f7e95ce07cccc4b64ebaba/clang-tools-extra/clangd/tool/ClangdMain.cpp#L397-L407)?
I can't imagine, clangd doesn't even re-use clang's PCHs.

**Comment by sam-mccall** [comment]
I think this is clangd doing a preamble build and then clangd reusing that AST for parsing the main file, with no PCH serialization in between.

I think there are significant technical obstacles (i.e. clang's current infra doesn't support this, we'd need to build a bunch of stuff):
 - nodes are not standalone, but coupled to an ASTContext. Reusing the preamble ASTContext for the main parse mutates it, which we can't undo in order to reuse it again.
 - parsing the main file actually mutates the referenced nodes (e.g. marking them "used", adding redecl chains, adding instantiations to templates). Again there's no mechanism to reverse this.
 - AST nodes and ASTContext are not threadsafe, and we rely on being able to concurrently parse with the same preamble (e.g. for code completion)
 - there is ASTImporter for cloning nodes across contexts. No idea what its performance is like. There are several other things preambles do that would need to be solved, e.g. lazy loading (cloning the whole AST will certainly be too slow).
 - preambles are a smaller representation than in-memory AST nodes, IIRC by a factor of around 3x. So memory usage would be much higher than even the preambles-in-memory option today.

Finding ways to avoid serialization cost would be a big deal, nobody is currently pursuing this. (I think serialization is more important than deserialization because we must serialize the whole AST but only deserialize what we need, albeit repeatedly).

Another thing that might be lower-hanging is not indexing the preamble AST synchronously before serializing it, but deferring it until the serialized preamble has been produced and can be used. This reduces the preamble->main AST->diagnostics critical path: it's typically ~1/3 indexing by CPU and ~1/6 indexing by wall-time in our environment.
